Enhance extension documentation (#11801)
gardener/gardener: docs/extensions/overview.md

Initially, everything was developed in-tree in the Gardener project. All cloud providers and the configuration for all the supported operating systems were released together with the Gardener core itself.
But as the project grew, it got more and more difficult to add new providers and maintain the existing code base.
As a consequence and in order to become agile and flexible again, we proposed [GEP-1](../proposals/01-extensibility.md) (Gardener Enhancement Proposal) and later [gardener/gardener#9635](https://github.com/gardener/gardener/issues/9635) as an enhancement.
The document describes an out-of-tree extension architecture that keeps the Gardener core logic independent of provider-specific knowledge (similar to what Kubernetes has achieved with [out-of-tree cloud providers](https://github.com/kubernetes/enhancements/issues/88) or with [CSI volume plugins](https://github.com/kubernetes/community/pull/1258)).

## Basic Concepts

> Gardener keeps running in the "garden cluster" and implements the core logic of garden, seed and shoot cluster reconciliation / deletion.

Reviewer: Here Gardener is a vague term that i guess means `gardener-operator` ( `Gardener keeps running in the "garden cluster" ...`). But it only reconciles the garden and not seeds and shoots ( `and implements the core logic of garden, seed and shoot cluster reconciliation / deletion.`)

Extensions are Kubernetes controllers themselves (like Gardener) and run in the garden runtime and seed clusters.
As usual, we try to use Kubernetes wherever applicable.
We rely on Kubernetes extension concepts in order to enable extensibility for Gardener.

### Building Blocks

Extensions consist of the following building blocks:
1. A Helm chart as the vehicle to generally deploy extension controllers to a Kubernetes clusters

> 1. Controllers that reconcile objects of the API group `extensions.gardener.cloud`. These controllers take over outsourced tasks, like creating the shoot infrastructure or deploying components to the control-plane. Optionally, extensions can bring their own webhooks to mutate resources deployed by Gardener.

Reviewer: To use the same term as the point above

```suggestion
1. Extension controllers that reconcile objects of the API group `extensions.gardener.cloud`. These controllers take over outsourced tasks, like creating the shoot infrastructure or deploying components to the control-plane. Optionally, extensions can bring their own webhooks to mutate resources deployed by Gardener.
```

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: Infrastructure
metadata:
  name: infrastructure
  namespace: shoot--core--aws-01
spec:
  type: local
```

### Classes

A Gardener landscape consists of various cluster types which extension controllers may consider during reconciliation.
The `.spec.class` field identifies the different deployment cases. 

#### Garden

Extensions serve the garden (run in garden runtime), e.g. installing certifictes for API and ingress endpoints.
In the course of the `Garden` reconciliation, the `gardener-operator` creates `BackupBucket`, `DNSRecord` and `Extension` resources (group `extensions.gardener.cloud`) which triggers the reconciliation of responsible extension controllers.

#### Seed

Extensions serve the seed (run in seed), e.g. requesting a wildcard certificate for the seed's ingress domain.
In the course of the `Seed` reconciliation, the `gardenlet` creates `DNSRecord` and `Extension` resources (group `extensions.gardener.cloud`) which triggers the reconciliation of responsible extension controllers.

#### Shoot

> Extensions serve the shoot (run in seed), e.g. deploying a certificate controller into the control-plane namespace.

Reviewer: Extensions is a little vague: Isn't this better? Suggestion for change in multiple places
```suggestion
Extension controllers serving the shoot (run in seed), e.g. deploying a certificate controller into the control-plane namespace.
```

> In the course of the `Shoot` reconciliation, the `gardenlet` creates various extension resources (group `extensions.gardener.cloud`) which triggers the reconciliation of responsible extension controllers.

Reviewer: Suggestion for change in multiple places
```suggestion
In the course of the `Shoot` reconciliation, the `gardenlet` creates various extension resources (group `extensions.gardener.cloud`) which triggers the responsible extension controllers to reconcile them
```
